File: solutions/java/src/main/java/com/epi/ClosestToMedian.java
package com.epi;

import java.util.*;

public class ClosestToMedian {

  // @include
  public static int[] findKClosestToMedian(int[] A, int k) {
    final double median = findMedian(A);

    // We will use this custom comparator to reorder the array
    // so that elements close to the median come to the front
    // of the array.
    Comparator<Integer> cmp = new Comparator<Integer>() {
        // Compare a and b based on how close they are to median.
        public int compare(Integer a, Integer b) {
            // Note the use of closure to access the variable median, which
            // is in the outer scope --- closure greatly simplifies the code.
            return Double.compare(Math.abs(a - median), Math.abs(b - median));
        }
    };

    OrderStatistic.findKth(A, k - 1, cmp);

    // Since findKth reordered A so that elements closest in absolute value
    // to median have been moved to the front of A, the first k entries 
    // are the result.
    return Arrays.copyOfRange(A, 0, k);
  }

  // Promote the return value to double to prevent precision error.
  public static double findMedian(int[] A) {
    int half = A.length / 2;
    OrderStatistic.findKthSmallest(A, half);

    if ((A.length % 2) != 0) { // A has an odd number elements.
      return A[half];
    } else { // A has an even number elements.
      int x = A[half];
      OrderStatistic.findKthSmallest(A, half - 1);
      return 0.5 * (x + A[half - 1]);
    }
  }
  // @exclude

  private static void checkAns(int[] answer, int[] res, int k) {
    Arrays.sort(answer);
    double median = ((answer.length & 1) != 0)
                        ? answer[(answer.length / 2)]
                        : 0.5 * (answer[((answer.length / 2) - 1)] +
                                 answer[(answer.length / 2)]);
    List<Double> temp = new ArrayList<>();
    for (int a : answer) {
      temp.add(Math.abs(median - a));
    }
    Collections.sort(temp);
    for (int r : res) {
      assert(Math.abs(r - median) <= temp.get(k - 1));
    }
  }

  private static void simpleTest() {
    int[] d = new int[]{3, 2, 3, 5, 7, 3, 1};
    int[] dExpres = new int[]{2, 3, 3};
    int[] dRes = findKClosestToMedian(d, 3);
    checkAns(d, dRes, 3);
    for (int dResElement : dRes) {
      System.out.print(dResElement + " ");
    }
    System.out.println();
  }

  static int[] toIntArray(List<Integer> list){
    int[] result = new int[list.size()];
    for(int i = 0; i < result.length; i++) {
      result[i] = list.get(i);
    }
    return result;
  }

  public static void main(String[] args) {
    Random r = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n, k;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
        k = r.nextInt(n) + 1;
      } else if (args.length == 2) {
        n = Integer.parseInt(args[0]);
        k = Integer.parseInt(args[1]);
      } else {
        n = r.nextInt(100000) + 1;
        k = r.nextInt(n) + 1;
      }
      List<Integer> a = new ArrayList<>();
      for (int i = 0; i < n; ++i) {
        a.add(r.nextInt(n * 2));
      }
      // System.out.println(a);
      int[] res = findKClosestToMedian(toIntArray(a), k);
      assert(res.length == k);
      System.out.println("n = " + n + ", k = " + k);
      checkAns(toIntArray(a), res, k);
    }
  }
}

File: solutions/cpp/Closest_to_median.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::nth_element;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

double FindMedian(vector<int>* A);

// @include
class Comp {
 public:
  explicit Comp(double m) : m_(m) {};

  bool operator()(int a, int b) const { return fabs(a - m_) < fabs(b - m_); }

 private:
  double m_;
};

vector<int> FindKClosestToMedian(vector<int> A, int k) {
  // Find the element i where |A[i] - median| is the k-th smallest.
  nth_element(A.begin(), A.begin() + k - 1, A.end(), Comp{FindMedian(&A)});
  return {A.cbegin(), A.cbegin() + k};
}

// Promote the return value to double to prevent precision error.
double FindMedian(vector<int>* A) {
  int half = A->size() / 2;
  nth_element(A->begin(), A->begin() + half, A->end());
  if (A->size() & 1) {  // A has odd number of elements.
    return (*A)[half];
  } else {  // A has even number of elements.
    int x = (*A)[half];
    nth_element(A->begin(), A->begin() + half - 1, A->end());
    return 0.5 * (x + (*A)[half - 1]);
  }
}
// @exclude

void CheckAns(vector<int>& A, const vector<int>& res, int k) {
  sort(A.begin(), A.end());
  double median = (A.size() & 1)
                      ? A[A.size() / 2]
                      : 0.5 * (A[(A.size() / 2) - 1] + A[A.size() / 2]);
  vector<double> temp;
  for (int a : A) {
    temp.emplace_back(fabs(median - a));
  }
  sort(temp.begin(), temp.end());
  for (int r : res) {
    assert(fabs(r - median) <= temp[k - 1]);
  }
}

void SimpleTest() {
  vector<int> D = {3, 2, 3, 5, 7, 3, 1};
  vector<int> Dexpres = {2, 3, 3};
  vector<int> Dres = FindKClosestToMedian(D, 3);
  CheckAns(D, Dres, 3);
  for (int d : Dres) {
    cout << d << ' ';
  }
  cout << endl;
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n, k;
    if (argc == 2) {
      n = atoi(argv[1]);
      uniform_int_distribution<int> k_dis(1, n);
      k = k_dis(gen);
    } else if (argc == 3) {
      n = atoi(argv[1]);
      k = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> n_dis(1, 100000);
      n = n_dis(gen);
      uniform_int_distribution<int> k_dis(1, n);
      k = k_dis(gen);
    }
    vector<int> A;
    uniform_int_distribution<int> dis(0, (n * 2) - 1);
    for (int i = 0; i < n; ++i) {
      A.emplace_back(dis(gen));
    }
    /*
    for (int a : A) {
      cout << a << ' ';
    }
    cout << endl;
    */
    vector<int> res = FindKClosestToMedian(A, k);
    assert(res.size() == k);
    cout << "n = " << n << ", k = " << k << endl;
    /*
    for (int a : res) {
      cout << a << ' ';
    }
    cout << endl;
    */
    CheckAns(A, res, k);
  }
  return 0;
}
